kafka-connect-aws-s3/src/main/scala/io/lenses/streamreactor/connect/aws/s3/sink/S3SinkTask.scala

```diff
@@ -41,44 +44,37 @@ import scala.util.Try
 
 class S3SinkTask extends SinkTask with ErrorHandler {
 
+  implicit val contextPropsFn: () => util.Map[String, String] = SinkContextReader.getProps(() => context)
+
   private val manifest = JarManifest(getClass.getProtectionDomain.getCodeSource.getLocation)
 
   private var writerManager: S3WriterManager = _
 
-  private var sinkName: String = _
+  private implicit var connectorTaskId: ConnectorTaskId = DefaultConnectorTaskId
```

Comment:
Unfortunately we are in a Java API that requires state here.  The `start` method has to set up the state based on the properties.  I can't think of a way to avoid these.

To avoid the mutability in the past I have introduced finite state models into the code.  This doesn't completely remove the mutability as you still have to have a variable to keep your state in.  If you have any ideas I'd be happy to hear them.

Reply:
From a concurrency perspective we have one writer : the framework itself calling start. Start won't be called more than once per task instance. Since it is a java API and the task needs to track its own state, we cannot avoid the var.